Test 8.2.1: Caff node reorg handling for unfinalized block

shenkeyao opened the pull request:
Closes https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1210039234519500?focus=true.

### This PR:
* Adds a test to verify the Caff node properly waits for the L1 finality.

### This PR does not:
* Test the sad path, i.e., the reorg handling for blocks that should be dropped. This will be covered by https://app.asana.com/1/1208976916964769/project/1209392461754458/task/1210127882718167?focus=true.

### How to test this PR:
* `go test ./espresso/environment/8_reorg_test.go -v` to run all tests in the file.

<!-- ### Things tested -->
<!-- Anything that was manually tested (that is not tested in CI). -->
<!-- E.g. building/running of docker containers. Changes to docker demo, ... -->
<!-- Especially mention anything untested, with reasoning and link an issue to resolve this. -->

<!-- Complete the following items before creating this PR -->
<!-- [ ] Issue linked or PR description mentions why this change is necessary. -->
<!-- [ ] PR description is clear enough for reviewers. -->
<!-- [ ] Documentation for changes (additions) has been updated (added).  -->
<!-- [ ] If this is a draft it is marked as "draft".  -->

<!-- To make changes to this template edit https://github.com/EspressoSystems/.github/blob/main/PULL_REQUEST_TEMPLATE.md -->


---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1210039234519499

shenkeyao committed:
Add test

shenkeyao committed:
Merge branch 'celo-integration' into keyao/caff-node-wait-finality

shenkeyao committed:
Fix the test

shenkeyao committed:
Merge branch 'keyao/test-batcher-unfinalized' into keyao/caff-node-wait-finality

shenkeyao committed:
Merge branch 'keyao/test-batcher-unfinalized' into keyao/caff-node-wait-finality

Ayiga commented on op-node/node/node.go:
```diff
@@ -554,6 +556,10 @@ func (n *OpNode) initP2PSigner(ctx context.Context, cfg *Config) (err error) {
 }
 
+func (n *OpNode) EspressoStreamer() *espresso.EspressoStreamer[derive.EspressoBatch] {
+	return n.l2Driver.SyncDeriver.Derivation.EspressoStreamer()
```
I like the idea here... But I think we need to be careful.  I believe this is a `Getter` and does not actually create a new instance.  Which means, if you utilize this call, and start consuming the streamer in the test, then you will be depriving the Caff Node itself of those same Batches, which may result in a stall.

philippecamacho replied:
I agree. I understand though that in this test it is not a problem as  we only we use  this Streamer instance to look at its batch buffer. In all cases we should comment this functionto highlight the risk mentioned by @Ayiga.

shenkeyao replied:
Added a comment to note this.

philippecamacho commented on espresso/environment/8_reorg_test.go:
```diff
@@ -86,3 +86,92 @@ func TestBatcherWaitForFinality(t *testing.T) {
+	defer cancel()
+
+	launcher := new(env.EspressoDevNodeLauncherDocker)
+	system, espressoDevNode, err := launcher.StartDevNet(ctx, t, 0)
```
```suggestion
	system, espressoDevNode, err := launcher.StartDevNet(ctx, t, 4)
```

As this test is about finality, I suggest to increase the `L1FinalizedDistance`. The test is still passing on my machine with `L1FinalizedDistance=4`.

shenkeyao replied:
Updated! Also updated the setting in the batcher test PR.

shenkeyao committed:
Merge branch 'keyao/test-batcher-unfinalized' into keyao/caff-node-wait-finality

shenkeyao committed:
Make test pass

shenkeyao committed:
Restore commented out test

shenkeyao committed:
Merge branch 'keyao/test-batcher-unfinalized' into keyao/caff-node-wait-finality

shenkeyao committed:
Make test pass deterministically

shenkeyao committed:
Merge branch 'keyao/test-batcher-unfinalized' into keyao/caff-node-wait-finality

Ayiga commented on espresso/environment/8_2_1_caff_node_wait_finality_test.go:
```diff
@@ -0,0 +1,94 @@
+	// Wait for the batch buffer to be empty which will trigger the Caff node to sync the status
+	// and insert more batches to the buffer.
+	for {
+		if streamer.BatchBuffer.Len() == 0 {
```
We are hot-looping here while checking for the length of the batch buffer.  If this code is run in a system with a single thread, we run the risk of potentially not making any progress, because this would be actively scheduled and no pause in the execution would prevent the scheduling of another goroutine.

Ultimately, the solution here is to add a `time.Sleep` before we loop again (especially if the length is not 0)

Ayiga commented on espresso/environment/8_2_1_caff_node_wait_finality_test.go:
```diff
@@ -0,0 +1,94 @@
+					if statusAfterWait.FinalizedL1.Number > initialStatus.FinalizedL1.Number {
+						// Verify that eventually the batch buffer will be updated.
+						tickerBufferInsert := time.NewTicker(100 * time.Millisecond)
+						defer tickerBufferInsert.Stop()
```
Ooo... `defer`s only evaluate / run when the entire function exits, not the closure.  So every time we hit this while we're iterating through this nested loop stack, we're potentially adding more and more defers without stopping the actual underlying Ticker.

This would be a non-issue if the contents of this conditional were in their own function.

shenkeyao replied:
Move it to a separate function!

Ayiga commented on espresso/environment/8_2_1_caff_node_wait_finality_test.go:
```diff
@@ -0,0 +1,94 @@
+			// Wait for the finalized L1 number and the batch buffer to be updated.
+			for {
+				if streamer.BatchBuffer.Len() > 0 {
+					// Verify that any batch inserted into the batch buffer has a finalized L1
```
The nesting of the for-loops and conditionals is getting quite deep.  We should separate this logic into separate functions to help readability and to help clarify the context we need to know for each section of logic.

shenkeyao replied:
Makes sense. Added two helper functions.

Ayiga commented on espresso/environment/8_2_1_caff_node_wait_finality_test.go:
```diff
@@ -0,0 +1,94 @@
+				if streamer.BatchBuffer.Len() > 0 {
+					// Verify that any batch inserted into the batch buffer has a finalized L1
+					// origin.
+					for {
```
So we want to loop while we iterate over all of the batches within our `BatchBuffer`.

Nit: I don't like that we're doing with the condition not being in the loop statement itself, but I understand the reason why.  Some potential alternatives, which are also less than ideal:

Option 1:
```go
batch := streamer.BatchBuffer.Pop()
for batch != nil {
    ...
    batch = streamer.BatchBuffer.Pop()
}
```

Option 2:
```go
for batch := streamer.BatchBuffer.Pop(); batch != nil; batch = streamer.BatchBuffer.Pop() {
    ...
}
```

Option 3:
```go
next := streamer.BatchBuffer.Pop
for batch := next(); batch != nil; batch = next() {
    ...
}
```

Option 4:
```go
next := streamer.BatchBuffer.Pop
batch := next()
for ; batch != nil; batch = next() {
    ...
}
```

shenkeyao replied:
Thanks for providing the options! I chose the first one. I still don't quite like the looping-while-updating implementation, but I think it's better than the previous version.

shenkeyao committed:
Merge branch 'keyao/test-batcher-unfinalized' into keyao/caff-node-wait-finality

shenkeyao committed:
Improve loops

shenkeyao committed:
Merge branch 'ag/l1reorg' into keyao/caff-node-wait-finality

shenkeyao commented:
@Ayiga @philippecamacho and whoever is going to re-review the changes: I had to rebase this branch off of @QuentinI's reorg test branch to include fixes to the Caff node/streamer, so before his PR is merged, this PR will contain many changes not made by me. For easier review, feel free to skip other files and only review `espresso/environment/8_2_1_caff_node_wait_finality_test.go` for now. I'll rebase again once that PR is merged!

shenkeyao committed:
Merge branch 'keyao/test-batcher-unfinalized' into keyao/caff-node-wait-finality

shenkeyao committed:
Move Caff node test to the reorg test file, fix the finalized number source

shenkeyao committed:
Merge branch 'celo-integration' into keyao/caff-node-wait-finality

philippecamacho commented on espresso/environment/8_reorg_test.go:
```diff
@@ -78,6 +79,109 @@ func TestBatcherWaitForFinality(t *testing.T) {
+//	Assert:
+//		The Caff node doesn't insert a batch without finalized L1 origin to the batch buffer.
+//		After the L1 origin is finalized, the Caff node inserts the batch.
+func TestCaffNodeWaitForFinality(t *testing.T) {
```
It seems to me that the other [test](https://github.com/EspressoSystems/optimism-espresso-integration/blob/bee7a4f643b2831ad5dbb34e1de769cfe4063f4a/espresso/environment/8_reorg_test.go#L30) was better in the sense that it was black box instead of relying on the inspection of the inner data structures of the Caff node. Let us try to understand why the [old test](https://github.com/EspressoSystems/optimism-espresso-integration/blob/bee7a4f643b2831ad5dbb34e1de769cfe4063f4a/espresso/environment/8_reorg_test.go#L30)  does not pass anymore. cc @QuentinI @Ayiga

shenkeyao replied:
Yeah would like to understand the differences better. Pasting my message from [Slack](https://espressosys.slack.com/archives/C08RHAAJU6S/p1747087808568349) so we could continue the discussion here:
> Did we have any recent change that might have caused the L1 finalized number from the `SyncStatus` to be inconsistent with the number from the streamer?
> The Caff node reorg test failed when fetching the L1 finalized number from `rollupClient.SyncStatus`, but passes after I updated the test to use the number from the Espresso streamer instead. In order to make the test pass, I think it makes sense to use the number from the streamer, because it uses the derivation pipeline as the source which should be accurate. But I’d like to understand the L1 finalized number from the `rollupClient` as well, so was wondering whether we changed anything that caused the `rollupClient` to be less up-to-date than before (or instead, did we change the streamer to be updated more often?). If so, we probably should avoid using the `rollupClient` as the status source in future tests as well.

shenkeyao replied:
Thanks for the team's debugging effort and @Ayiga's latest commit. This has been fixed.

Ayiga committed:
Adjust VerifyL1OriginFinalizedCheck to reference L1 finalized block

The `TestCaffNodeWaitForFinality` test was a bit flaky as it was utilizing
the L2Verifier's understanding of what the finalized L1 was.  This can be
problematic, and was causing the test to fail occasionally as it can be
behind reality.

To address this, we instead utilize the L1's understanding of what it's
finalized block is.  This is more in line with the spirit of the original
test's intention, and helps the test to pass more consistently.

philippecamacho reviewed (approved):
LGTM
